File: JavaR2/src/Solution0028.java
public class Solution0028 {
	public int strStr(String haystack, String needle) {
		if (needle.length() == 0) {
			return 0;
		}

		int[] next = getNext(needle);

		int j = 0;
		for (int i = 0; i < haystack.length(); i++) {
			while (j > 0 && haystack.charAt(i) != needle.charAt(j)) {
				j = next[j-1];
			}
			if (haystack.charAt(i) == needle.charAt(j)) {
				j++;
			}
			if (j == needle.length()) {
				return i - needle.length() + 1;
			}
		}

		return -1;
	}

	//计算前缀表
	public static int[] getNext(String t) {
		int len = t.length();
		int[] next = new int[len];
		int j = 0;

		//i指向后缀末尾， j指向前缀末尾
		for (int i = 1; i < t.length(); i++) {
			while (j > 0 && t.charAt(i) != t.charAt(j)) {
				j = next[j-1];
			}
			if (t.charAt(i) == t.charAt(j)) {
				j++;
			}
			next[i] = j;
		}
		return next;
	}
}

File: goR3/solution0028.go
package main

func StrStr(haystack string, needle string) int {
	//滑动窗口
	i, j, k := 0, 0, 0

	for i < len(haystack) && j < len(haystack) {
		if i == j {
			//滑动窗口开始，移动j
			if haystack[i] != needle[0] {
				i++
				j++
			} else {
				j++
				k++
			}
		}
		if i != j {
			for j < len(haystack) && j-i < len(needle) &&
				haystack[j] == needle[k] {
				j++
				k++
			}
			if k == len(needle) {
				return i
			} else {
				i++
				j = i
				k = 0
			}
		}
	}
	return -1
}
